### PR TITLE
Add tests for gRPC client-side cancellation

### PR DESCRIPTION
**Pull request by tanmayv25**

### CODE REVIEW

File: qa/L0_client_cancellation/test.sh [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+fi
+if [ ! -z "$TEST_REPO_ARCH" ]; then
+    REPO_VERSION=${REPO_VERSION}_${TEST_REPO_ARCH}
+fi
+
+export CUDA_VISIBLE_DEVICES=0
+
+RET=0
+
+CLIENT_CANCELLATION_TEST=client_cancellation_test.py
```
Comment (rmccorm4) [inline qa/L0_client_cancellation/test.sh:45]: You and @Tabrizian should probably sync up on a test name/folder, I think he is [adding tests as well](https://github.com/triton-inference-server/server/pull/6252) and it probably makes sense for them to be in the same QA folder, or have a consistent prefix if separate folders.

File: qa/L0_client_cancellation/client_cancellation_test.py [hunk 2]
```diff
@@ -0,0 +1,232 @@
+#!/usr/bin/env python3
+
+# Copyright 2020-2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```
Comment (rmccorm4) [inline qa/L0_client_cancellation/client_cancellation_test.py:3]:
```suggestion
# Copyright 2023, NVIDIA CORPORATION & AFFILIATES. All rights reserved.
```

File: qa/L0_client_cancellation/client_cancellation_test.py [hunk 3]
```diff
@@ -54,6 +54,7 @@ def callback(user_data, result, error):
     else:
         user_data._completed_requests.put(result)
 
+
```
Comment (jbkyang-nvi) [inline qa/L0_client_cancellation/client_cancellation_test.py:57]: nit: consistent spacing between functions? Or is this something the autoformatter did

File: qa/L0_client_cancellation/client_cancellation_test.py [hunk 4]
```diff
@@ -0,0 +1,237 @@
...
+import socket
+import unittest
+from functools import partial
+import time
+
+import numpy as np
+import test_util as tu
+import tritonclient.grpc as grpcclient
+import tritonclient.grpc.aio as aiogrpcclient
+from tritonclient.utils import InferenceServerException
```
[the remark this answers is not in the document]
Comment (jbkyang-nvi) [reply qa/L0_client_cancellation/client_cancellation_test.py:43]: ^

File: qa/L0_client_cancellation/client_cancellation_test.py [hunk 5]
```diff
@@ -0,0 +1,235 @@
+#!/usr/bin/env python3
```
Comment (Tabrizian) [inline qa/L0_client_cancellation/client_cancellation_test.py:1]: Move the test to `L0_request_cancellation`.

File: qa/L0_request_cancellation/models/custom_identity_int32/config.pbtxt [hunk 6]
```diff
@@ -0,0 +1,54 @@
+# Copyright (c) 2020-2021, NVIDIA CORPORATION. All rights reserved.
```
Comment (Tabrizian) [inline qa/L0_request_cancellation/models/custom_identity_int32/config.pbtxt:1]: Update copyright

File: Dockerfile.QA [hunk 7]
```diff
@@ -139,6 +139,7 @@ RUN mkdir -p qa/common && \
     cp bin/data_compressor_test qa/L0_data_compression/. && \
     cp bin/metrics_api_test qa/L0_metrics/. && \
     cp bin/response_cache_test qa/L0_response_cache/. && \
+    cp bin/request_cancellation_test qa/L0_request_cancellation/. && \
```
Comment (Tabrizian) [inline Dockerfile.QA:142]: These changes are included in the side branch. Is the branch rebased?

File: qa/L0_request_cancellation/client_cancellation_test.py [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+            tasks.append(asyncio.create_task(handle_response(generator)))
+            tasks.append(asyncio.create_task(cancel_request(grpc_call)))
+
+            for task in tasks:
+                await task
+
+            self._record_end_time_ms()
+            self._test_runtime_duration(5000)
+
+        asyncio.run(test_aio_infer(self))
```
Comment (Tabrizian) [inline qa/L0_request_cancellation/client_cancellation_test.py:194]: I'm fine with we have but you could also use the `unittest.IsolatedAsyncioTestCase` for unit testing async methods. You can use this as a reference: 

https://github.com/triton-inference-server/server/blob/7624490df45549a2aa4f2ed1d5c963414a265c62/qa/L0_grpc/python_grpc_aio_test.py#L34C25-L34C57

File: qa/L0_request_cancellation/test.sh [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+rm -f *.log.*
+
+CLIENT_LOG=`pwd`/client.log
+DATADIR=`pwd`/models
+SERVER=/opt/tritonserver/bin/tritonserver
+SERVER_ARGS="--model-repository=$DATADIR --log-verbose=1"
+source ../common/util.sh
+
+SERVER_LOG=server.log
+LD_LIBRARY_PATH=/opt/tritonserver/lib:$LD_LIBRARY_PATH ./request_cancellation_test > $SERVER_LOG
```
Comment (Tabrizian) [inline qa/L0_request_cancellation/test.sh:64]: Needs rebasing. It looks like `test.sh` is created from scratch in here.

File: qa/L0_request_cancellation/client_cancellation_test.py (hunk 8 above)
Comment (tanmayv25) [reply qa/L0_request_cancellation/client_cancellation_test.py]: I have deferred adding asyncio request cancellation to a later point. We can explore better ways to implement and test asyncio library

**Review by kthui — APPROVED** [review]

**Pull request merged**